internal/database/database.go

```diff
@@ -409,10 +433,64 @@ func (d *cosmosDBClient) GetResourceDoc(ctx context.Context, resourceID *azcorea
 	return typedDoc.ID, innerDoc, nil
 }
 
+func (d *cosmosDBClient) GetHCPCluster(ctx context.Context, resourceID *azcorearm.ResourceID) (string, *HCPClusterDocument, error) {
+	typedDoc, err := d.getResourceDoc(ctx, resourceID)
+	if err != nil {
+		return "", nil, err
+	}
+
+	ret := &HCPClusterDocument{}
+	err = json.Unmarshal(typedDoc.Properties, ret)
+	if err != nil {
+		return "", nil, err
+	}
+
+	// TODO make all the typed doc stuff public
+	ret.TypedDocument = TypedDocument{
+		BaseDocument: BaseDocument{
+			ID:                typedDoc.ID,
+			TimeToLive:        typedDoc.TimeToLive,
+			CosmosResourceID:  typedDoc.CosmosResourceID,
+			CosmosSelf:        typedDoc.CosmosSelf,
+			CosmosETag:        typedDoc.CosmosETag,
+			CosmosAttachments: typedDoc.CosmosAttachments,
+			CosmosTimestamp:   typedDoc.CosmosTimestamp,
+		},
+		PartitionKey: typedDoc.PartitionKey,
+		ResourceType: typedDoc.ResourceType,
+	}
+
+	if ret.ResourceID.ResourceType.String() != api.ClusterResourceType.String() {
+		return "", nil, &typedDocumentError{
+			invalidType:    typedDoc.ResourceType,
+			propertiesType: "HCPClusterDocument",
+		}
+
+	}
+
+	// Replace the key field from Cosmos with the given resourceID,
+	// which typically comes from the URL. This helps preserve the
+	// casing of the resource group and resource name from the URL
+	// to meet RPC requirements:
+	//
+	// Put Resource | Arguments
+	//
+	// The resource group names and resource names should be matched
+	// case insensitively. ... Additionally, the Resource Provier must
+	// preserve the casing provided by the user. The service must return
+	// the most recently specified casing to the client and must not
+	// normalize or return a toupper or tolower form of the resource
+	// group or resource name. The resource group name and resource
+	// name must come from the URL and not the request body.
+	ret.ResourceID = resourceID
+
+	return typedDoc.ID, ret, nil
+}
+
 func (d *cosmosDBClient) CreateResourceDoc(ctx context.Context, doc *ResourceDocument) error {
 	typedDoc := newTypedDocument(doc.ResourceID.SubscriptionID, doc.ResourceID.ResourceType)
 
-	data, err := typedDocumentMarshal(typedDoc, doc)
+	data, err := typedDocumentMarshal(typedDoc, doc, ResourceDocument{}.GetValidTypes()...)
```

Comment:
Once we have separate schema types, we won't have to worry about the serialization being incorrect.